<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Chaining Promises</title>
  </head>
  <body>
    <script>
    const posts = [
      { title: 'I love JavaScript', author: 'Wes Bos', id: 1 },
      { title: 'CSS!', author: 'Chris Coyierr', id: 2 },
      { title: 'Dev tools tricks', author: 'Addy Osmani', id: 3 },
    ];

    const authors = [
      { name: 'Wes Bos', twitter: '@wesbos', bio: 'Canadian Developer' },
      { name: 'Chris Coyier', twitter: '@chriscoyier', bio: 'CSS Tricks and CodePen' },
      { name: 'Addy Osmani', twitter: '@addyosmani', bio: 'Googler' },
    ];

    // This function creates a new promise when looking for a post from the data, if successful it returns the post otherwise gives an error message
    function getPostById(id) {
      const post = posts.find(post => post.id == id);
      return new Promise((resolve, reject) => {
        post ? resolve(post) : reject(Error('No Post Was Found!'));
      })
    }

    // This function finds an author for the post object returned from the previous promise if it exists, the author is replaced by the corresponding author object
    function hydrateAuthor(post) {
      const authorInfo = authors.find(author => author.name == post.author);
      return new Promise((resolve, reject) => {
        if (authorInfo) {
          post.author = authorInfo;
          resolve(post);
        } else {
          reject(Error('No Author For This Post'))
        }
      })
    }

    // If returning a promise in a .then() method, you can chain them
    getPostById(2).then(post =>
      {
        return hydrateAuthor(post);
      }).then(post =>
        {
          console.log(post)
        }
        // To display promise errors use the .catch() method
      ).catch(err =>
        console.error(err)
      )

    </script>
  </body>
</html>
